platform/kube/admit/admit.go

```diff
@@ -0,0 +1,248 @@
+// limitations under the License.
+
+package admit
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+	"k8s.io/api/admission/v1alpha1"
+	admissionregistrationv1alpha1 "k8s.io/api/admissionregistration/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/client-go/kubernetes"
+	admissionClient "k8s.io/client-go/kubernetes/typed/admissionregistration/v1alpha1"
+
+	"istio.io/pilot/adapter/config/crd"
+	"istio.io/pilot/model"
+)
+
+// A cluster-unique (i.e. random) suffix should be added to each
+// default below when testing in a shared cluster to avoid collisions.
+
+const (
+	// DefaultAdmissionHookConfigName is the default name for the
+	// ExternalAdmissionHookConfiguration.
+	DefaultAdmissionHookConfigName = "pilot-config"
+
+	// DefaultAdmissionHookName is the default name for the
+	// ExternalAdmissionHooks.
+	DefaultAdmissionHookName = "pilot.config.istio.io"
+
+	// DefaultAdmissionServiceName is the default service of the
+	// validation webhook.
+	DefaultAdmissionServiceName = "istio-pilot-config"
+)
+
+// Admit implements the external admission webhook for validation
+// pilot configuration.
+type Admit struct {
+	descriptor       model.ConfigDescriptor
+	hookConfigName   string
+	hookName         string
+	serviceName      string
+	serviceNamespace string
+	caBundle         []byte
+
+	// unconditionally validate all config that is not in this list of
+	// configuration.
+	validateNamespaces []string
+}
+
+// GetAPIServerExtensionCACert gets the CA cert that will signed the
+// cert used by the "GenericAdmissionWebhook" plugin admission
+// controller.
+func GetAPIServerExtensionCACert(cl kubernetes.Interface) ([]byte, error) {
+	const name = "extension-apiserver-authentication"
+	c, err := cl.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	pem, ok := c.Data["requestheader-client-ca-file"]
+	if !ok {
+		return nil, fmt.Errorf("cannot find ca.crt in %v: ConfigMap.Data is %#v", name, c.Data)
+	}
+	return []byte(pem), nil
+}
+
+// MakeTLSConfig makes a TLS configuration suitable for use with the
+// GenericAdmissionWebhook.
+func MakeTLSConfig(serverCert, serverKey, caCert []byte) (*tls.Config, error) {
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+	cert, err := tls.X509KeyPair(serverCert, serverKey)
+	if err != nil {
+		return nil, err
+	}
+	return &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		ClientCAs:    caCertPool,
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+	}, nil
+}
+
+// New creates a new instance of the admission webhook controller.
+func New(descriptor model.ConfigDescriptor, hookConfigName, hookName, serviceName, serviceNamespace string, validateNamespaces []string, caBundle []byte) *Admit { // nolint: lll
```

Comment:
nit: constructors are usually called `Create` or `Make` around here. Too bad there's no standard name for the pattern. Also, `Admit` does not sound good as a thing, maybe call it `AdmissionController`.